image_view/src/image_view_node.cpp

```diff
@@ -0,0 +1,254 @@
+  cv_bridge::CvImageConstPtr cv_ptr;
+
+  // Convert to OpenCV native BGR color
+  try {
+    cv_bridge::CvtColorForDisplayOptions options;
+    options.do_dynamic_scaling = do_dynamic_scaling;
+    options.colormap = g_colormap;
+    // Set min/max value for scaling to visualize depth/float image.
+    if (g_min_image_value == g_max_image_value) {
+      // Not specified by rosparam, then set default value.
+      // Because of current sensor limitation, we use 10m as default of max range of depth
+      // with consistency to the configuration in rqt_image_view.
+      options.min_image_value = 0;
+      if (msg->encoding == "32FC1") {
+        options.max_image_value = 10;  // 10 [m]
+      } else if (msg->encoding == "16UC1") {
+        options.max_image_value = 10 * 1000;  // 10 * 1000 [mm]
+      }
+    } else {
+      options.min_image_value = g_min_image_value;
+      options.max_image_value = g_max_image_value;
+    }
+
+    cv_ptr = cv_bridge::cvtColorForDisplay(cv_bridge::toCvShare(msg), "", options);
+    g_last_image = cv_ptr->image;
+  } catch (cv_bridge::Exception& e) {
+    RCLCPP_ERROR_EXPRESSION(
+      this->get_logger(), (static_cast<int>(this->now().seconds()) % 30 == 0),
+      "Unable to convert '%s' image for display: '%s'",
+      msg->encoding.c_str(), e.what());
+  }
+
+  if (g_gui && !g_last_image.empty()) {
+    const cv::Mat &image = g_last_image;
+    cv::imshow(window_name_, image);
+    cv::waitKey(1);
+  }
+
+  if (pub_->get_subscription_count() > 0) {
+    pub_->publish(*(cv_ptr->toImageMsg()));
+  }
+}
+
+void ImageViewNode::guiCb()
+{
+  // Process pending GUI events and return immediately
+  cv::waitKey(1);
+}
+
+void ImageViewNode::mouseCb(int event, int x, int y, int flags, void * param)
+{
+  (void)x;
+  (void)y;
+  (void)flags;
+
+  ImageViewNode *this_ = reinterpret_cast<ImageViewNode *>(param);
+
+  if (event == cv::EVENT_LBUTTONDOWN) {
+    RCLCPP_WARN_ONCE(this_->get_logger(), "Left-clicking no longer saves images. Right-click instead.");
+    return;
+  }
+
+  if (event != cv::EVENT_RBUTTONDOWN) {
+    return;
+  }
+  
+  cv::Mat image(this_->shown_image_.get());
+
+  if (image.empty()) {
+    RCLCPP_WARN(this_->get_logger(), "Couldn't save image, no data!");
+    return;
+  }
+
+  std::string filename = (this_->filename_format_ % this_->count_).str();
+
+  if (cv::imwrite(filename, image)) {
+    RCLCPP_INFO(this_->get_logger(), "Saved image %s", filename.c_str());
+    this_->count_++;
+  } else {
+    /// @todo Show full path, ask if user has permission to write there
+    RCLCPP_ERROR(this_->get_logger(), "Failed to save image.");
+  }
+}
+
+void ImageViewNode::windowThread()
+{
+  cv::namedWindow(window_name_, autosize_ ? cv::WND_PROP_AUTOSIZE : 0);
+  cv::setMouseCallback(window_name_, &ImageViewNode::mouseCb, this);
+
+  if(!autosize_) {
+    int width = this->declare_parameter("width", -1);
+    int height = this->declare_parameter("height", -1);
+
+    if (width > -1 && height > -1) {
```

Comment:
You can also declare and get, then check if its not been declared, but this is also OK